### ISSUE TITLE
Solaris 11 fails for string-to-key-test

### ISSUE THREAD

**Issue by quanah** [heimdal/heimdal#316]
```
string_to_key(nisse@FOO.SE, hej) failed
should be: 0ebc239d6846f2d551985b57c157017904c4e9fec10e13d0
result was: 0ebc239d6846f2d55164a4a42c2acdbcec1c64a4a42c2a7f
string_to_key(assar/liten@FOO.SE, hemligt) failed
should be: 7f4067b9bcc440fb4373d9d3cd7cc767e67994d0a834df62
result was: 7f4067b9bcc440fb437049c18a9b3bcbae7a7049c18a9b94
string_to_key(raeburn@ATHENA.MIT.EDU, password) failed
should be: 850bb51358548cd05e86768c313e3bfef7511937dcf72c3e
result was: 850bb51358548cd05e43baa22ae3b583f89e43baa22ae301
string_to_key(danny@WHITEHOUSE.GOV, potatoe) failed
should be: dfcd233dd0a43204ea6dc437fb15e061b02979c1f74f377a
result was: dfcd233dd0a43204ea025708436813ec04640257084368d9
string_to_key(buckaroo@EXAMPLE.COM, penny) failed
should be: 6d2fcdf2d6fbbc3ddcadb5da5710a23489b0d3b69d5d9d4a
result was: 6d2fcdf2d6fbbc3ddcf7da511cecd0b9f76ef7da511cec76
string_to_key(Jurišić@ATHENA.MIT.EDU, ß) failed
should be: 16d5a40e1ce3bacb61b9dce00470324c831973a7b952feb0
result was: 16d5a40e1ce3bacb61ae31736885feb06746ae3173688567
FAIL: string-to-key-test
```

**Comment by sfhacker** [comment]
Just built master branch from source and got this:

PASS string-to-key-test.exe (exit status: 0)

**Comment by SethEllsworth** [comment]
I get the exact same failures trying on both Solaris 10 sparc and x64:
should be: 0ebc239d6846f2d551985b57c157017904c4e9fec10e13d0
result was: 0ebc239d6846f2d55164a4a42c2acdbcec1c64a4a42c2a7f
...

Every failing string seems to deviate after the 18th character.

**Comment by dariaphoebe** [comment]
Works when using hcrypto instead of pkcs11 as crypto provider in evp.c

**Comment by dariaphoebe** [comment]
@lhoward do you have ideas about this? pkcs11 code came from you

**Comment by lhoward** [comment]
I’ll have to check out – so it’s failing when *not* using the pkcs11 backend?

**Comment by dariaphoebe** [comment]
It fails *only* when using the pkcs11 backend. The hcrypto backend is fine.

**Comment by lhoward** [comment]
Reproduced, investigating.

**Comment by lhoward** [comment]
Appears related to the fix introduced in #194.

**Comment by lhoward** [comment]
Reverting 81c778e fixes it, so perhaps doing that and revisiting #194 is the right fix.

**Comment by lhoward** [comment]
The #194 fix is definitely correct for test_crypto, results are identical between the hcrypto and PKCS#11 backends (across block sizes). But something strange is going on with the PKCS#11 backend when it comes to passing the AES and string2key tests, it seems that unless I back out #194 (which effectively resets the IV on each operation – degenerates to ECB mode if the input length is the block size), they fail. Still investigating.

**Comment by lhoward** [comment]
OK, I found the problem and have a fix; a cleaner one will come tomorrow. A clue: EVP_CIPH_ALWAYS_CALL_INIT.

**Comment by lhoward** [comment]
Needs a better commit message, but see 72d4f1be for now.

**Comment by lhoward** [comment]
(Please don't merge, more cleanups coming.)

**Comment by quanah** [comment]
With 2b5b7cf string-to-key-test passes for me

**Comment by lhoward** [comment]
Fixed in a88d0066.
